### PR TITLE
Feature/283 improve health endpoint with missing required conditions

### PR DESCRIPTION
**Pull request by berkaykrc**
## Description

This pull request introduces a health check system across multiple services (authentication, database, and song service) and implements error handling for service failures. It also includes logging enhancements, configuration fixes, and additional unit tests to ensure robustness.

## Checklist

- [x] I have read and followed the [Contributing Guidelines](https://github.com/AntonioMrtz/SpotifyElectron/blob/master/docs/CONTRIBUTING.md).
- [x] I have followed the project's [git convention](https://github.com/AntonioMrtz/SpotifyElectron/blob/master/docs/developer/Git-Convention.md) for the Pull Request's name and git commits.
- [x] I have tested the changes and verified they work as intended. Automated tests (unit, integration, e2e as applicable) and manual tests (UI interactions, edge cases, error states, etc.).
- [ ] I have set the pull request base to the appropriate `development-vx.xx` branch corresponding to the issue milestone.

## Commit type
feat

## Issue

#280 

### Health Check System Enhancements:
* Added a health check endpoint in `Backend/app/spotify_electron/health/health_controller.py` to validate the health of the database, authentication, and song services. The endpoint responds with appropriate HTTP status codes based on service health.
* Introduced `check_auth_health`, `check_database_health`, and `check_service_health` methods in their respective modules to verify service functionality and raise specific exceptions for failures. [[1]](diffhunk://#diff-cb569a18981cfe3fb704a2975d1361ed9b6d77eee4d7c71e994f8675c9a6b7d9R61-R87) [[2]](diffhunk://#diff-76b6fe6a2bd7ffacb39730f850e382a2b803018ec8dd9950431ae7e94fc0fcc2R62-R89) [[3]](diffhunk://#diff-4735a7017718680892a8b620be4cf811f0af8e5832d1f440c7e04f2373d15997R46-R79)

### Logging and Error Handling Improvements:
* Added logging for service health checks and database initialization in `Backend/app/auth/auth_schema.py`, `Backend/app/database/DatabaseConnectionManager.py`, and `Backend/app/spotify_electron/song/providers/song_service_provider.py`. [[1]](diffhunk://#diff-cb569a18981cfe3fb704a2975d1361ed9b6d77eee4d7c71e994f8675c9a6b7d9R61-R87) [[2]](diffhunk://#diff-76b6fe6a2bd7ffacb39730f850e382a2b803018ec8dd9950431ae7e94fc0fcc2R62-R89) [[3]](diffhunk://#diff-4735a7017718680892a8b620be4cf811f0af8e5832d1f440c7e04f2373d15997R46-R79)
* Introduced new error classes (`AuthServiceHealthCheckError`, `DatabasePingFailedError`, `SongServiceHealthCheckError`) for more granular exception handling. [[1]](diffhunk://#diff-cb569a18981cfe3fb704a2975d1361ed9b6d77eee4d7c71e994f8675c9a6b7d9R193-R201) [[2]](diffhunk://#diff-c2d52f7a05e273c36a395dfb325ae07163208f79e85dfc07b5ada73c0ccffedcR133-R153) [[3]](diffhunk://#diff-5b4d78bbb15b7af4d830ac553d85de9794ef20a1159782d1f6647f7137afe247R178-R186)

### Configuration Fixes:
* Fixed a typo in the `VERIFICATION_ALGORITHM` attribute in `Backend/app/common/app_schema.py` and its usage in `Backend/app/__main__.py`. [[1]](diffhunk://#diff-a8405e6e9d4fff3fe1cff710a4a08380c1b871d7139fa14e1112752278daf85bL57-R57) [[2]](diffhunk://#diff-7e39f8fc33922c4ba2c6b51a477aceabcc06dba99a93472dcbfa68f300c05816L58-R58)
* Corrected a logging constant typo in `Backend/app/logging/logging_constants.py`.

### Unit Testing Enhancements:
* Added comprehensive unit tests in `Backend/tests/test__health.py` to cover all possible health check scenarios, including healthy services, unhealthy services, and exceptions.

### Miscellaneous:
* Updated `Backend/app/resources/messages.ini` with new error messages for unhealthy services. [[1]](diffhunk://#diff-03c349aecf9675e1783bb001b8853ef5e23f2f0e8015f9c6d68e4a10bca2e275R28-R38) [[2]](diffhunk://#diff-03c349aecf9675e1783bb001b8853ef5e23f2f0e8015f9c6d68e4a10bca2e275R49-R51)
* Enhanced documentation in `Electron/src/swagger/api/services/HealthService.ts` to reflect the updated health check behavior.

## Assigned

@AntonioMrtz

### CODE REVIEW

File: Backend/app/spotify_electron/song/providers/song_service_provider.py [hunk 1]
```diff
@@ -42,6 +43,40 @@ def get_song_service(cls) -> ModuleType:
...
+                    cls.logger.warning(f"Song servise missing required method: {method}")
+        except Exception as exception:
+            cls.logger.exception("Error checking song service health")
+            raise SongServiceHealthCheckError from exception
+        else:
+            cls.logger.info("Song service health check successful")
+            return True
+
+    @classmethod
+    def _get_current_architecture(cls) -> AppArchitecture:
```
Comment (AntonioMrtz) [inline Backend/app/spotify_electron/song/providers/song_service_provider.py:76]: I like this

File: Backend/app/spotify_electron/song/providers/song_service_provider.py [hunk 2]
```diff
@@ -42,6 +43,40 @@ def get_song_service(cls) -> ModuleType:
         """
         return SongServiceProvider.song_service
 
+    @classmethod
+    def check_service_health(cls) -> bool:
+        """Check if the song service initialized and functioning.
+
+        Returns:
+            bool: True if the song service is operational, False otherwise
```
Comment (AntonioMrtz) [inline Backend/app/spotify_electron/song/providers/song_service_provider.py:51]: We switched from using `google` docstrings to `google-notypes` as we were not able to check if the type hints matched the ones specified in the docs. We won't use types in docstring in the future

File: Backend/app/resources/messages.ini [hunk 3]
```diff
@@ -42,3 +46,6 @@ token.invalid.credentials.auto.login = Invalid credentials provided during auto
 
 [GENRE]
 genre.not.valid = Genre not valid
+
+[DATABASE]
+database.unhealthy = Database unhealthy
```
Comment (AntonioMrtz) [inline Backend/app/resources/messages.ini:51]: `Database is unhealthy`

File: Backend/app/resources/messages.ini [hunk 4]
```diff
@@ -25,13 +25,17 @@ artist.not.found = Artist was not found
...
+auth.service.unhealthy = Auth service unhealthy
+
 [SONG]
 song.not.found = Song was not found
 song.data.not.found = Song data was not found
 song.bad.name = Song with invalid name
 song.bad.file = Song with invalid file
 song.already.exists = Song already exists
 song.create.unauthorized.user = Song cannot be created by a user
+song.service.unhealthy = Song services is unhealthy
```
Comment (AntonioMrtz) [inline Backend/app/resources/messages.ini:38]: `Song service is unhealthy`

File: Backend/app/resources/messages.ini [hunk 5]
```diff
@@ -25,13 +25,17 @@ artist.not.found = Artist was not found
 artist.bad.name = Artist with invalid name
 artist.already.exists = Artist already exists
 
+[AUTH]
+auth.service.unhealthy = Auth service unhealthy
```
Comment (AntonioMrtz) [inline Backend/app/resources/messages.ini:29]: `Auth service is unhealthy`

File: Backend/app/logging/logging_constants.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
 
 
 # Middleware
-LOGGIN_CHECK_AUTH_JWT_MIDDLEWARE = "CHECK_AUTH_JWT_MIDDLEWARE"
+LOGGING_CHECK_AUTH_JWT_MIDDLEWARE = "CHECK_AUTH_JWT_MIDDLEWARE"
```
Comment (AntonioMrtz) [inline Backend/app/logging/logging_constants.py:22]: Good catch

File: Backend/app/database/database_schema.py [hunk 7]
```diff
@@ -129,6 +130,27 @@ def __assert_connection_is_initialized(
...
+    @classmethod
+    async def check_connection_health(cls) -> bool | None:
+        """Check the health status of the database connection.
+
+        Raises:
+            DatabasePingFailedError: When the ping command fails or an error occurs while
+            communicating with the database
+
+        Returns:
+            bool: True if the database connection is healthy, False otherwise
```
Comment (AntonioMrtz) [inline Backend/app/database/database_schema.py:142]: No types on return

File: Backend/app/database/database_schema.py [hunk 8]
```diff
@@ -129,6 +130,27 @@ def __assert_connection_is_initialized(
     ):
         assert cls._connection is not None, "Database connection is not initialized"
 
+    @classmethod
+    async def check_connection_health(cls) -> bool | None:
```
Comment (AntonioMrtz) [inline Backend/app/database/database_schema.py:134]: Maybe we can use this method in the `init_connection` method

**Review by AntonioMrtz — CHANGES_REQUESTED** [review]

File: Backend/tests/test__health.py [hunk 9]
```diff
@@ -12,7 +16,105 @@ def set_up(trigger_app_startup):
...
+    mocker.patch(
+        "app.auth.auth_schema.AuthConfig.check_auth_health",
+        side_effect=mock_auth_service_exception,
+    )
+    response = health_check()
+    assert response.status_code == HTTP_503_SERVICE_UNAVAILABLE
+
+
+def test_health_check_multiple_services_unhealthy(mocker):
+    """Test health endpoint returns 500 when multiple services are unhealthy."""
```
Comment (AntonioMrtz) [inline Backend/tests/test__health.py:75]: This test docstring doesn't match the test behaviour

File: Backend/app/spotify_electron/song/providers/song_service_provider.py [hunk 10]
```diff
@@ -42,6 +44,23 @@ def get_song_service(cls) -> ModuleType:
...
+    def check_song_service_health(cls) -> None:
+        """Check if the song service initialized and functioning.
+
+        Raises:
+            SongServiceHealthCheckError: When health check fail on service provider
+        """
+        if not getattr(cls, "song_service", None):
+            cls.logger.warning("Song service not initialized")
+            raise SongServiceHealthCheckError
+        cls.logger.info("Song service health check successful")
```
Comment (AntonioMrtz) [inline Backend/app/spotify_electron/song/providers/song_service_provider.py:57]: I would use debug because of docker health check being called every 5s

File: Backend/app/auth/auth_schema.py [hunk 11]
```diff
@@ -56,11 +58,28 @@ def init_auth_config(
...
+            AuthServiceHealthCheckError: If the Auth configuration is not properly initialized.
+        """
+        if (
+            not hasattr(cls, "ACCESS_TOKEN_EXPIRE_MINUTES")
+            or not hasattr(cls, "DAYS_TO_EXPIRE_COOKIE")
+            or not hasattr(cls, "VERIFICATION_ALGORITHM")
+            or not hasattr(cls, "SIGNING_SECRET_KEY")
+        ):
+            raise AuthServiceHealthCheckError
+        cls.logger.info("Auth configuration health check successful")
```
Comment (AntonioMrtz) [inline Backend/app/auth/auth_schema.py:82]: I would use debug because of docker health check being called every 5s

File: Backend/app/database/DatabaseConnectionManager.py [hunk 12]
```diff
@@ -52,6 +59,23 @@ async def init_database_connection(
...
+        """Check if the database connection is established and working.
+
+        Raises:
+            DatabasePingFailedError:
+                When a connection error occurs while communicating with the database
+        """
+        cls.__assert_connection_is_initialized()
+        try:
+            await cls.__connection.check_connection_health()
+            cls.__logger.info("Database connection health check successful")
```
Comment (AntonioMrtz) [inline Backend/app/database/DatabaseConnectionManager.py:75]: I would use debug because of docker health check being called every 5s

**Review by AntonioMrtz — CHANGES_REQUESTED** [review]
The core of the PR looks good. Just a few little adjustments are needed.

In the future I will replace all `getattr`, `hasattr` methods with type safety access in PropertiesManager but this is out of the scope of this task

**Comment by AntonioMrtz** [comment]
Let me know when PR is ready to get a review again 😄

**Comment by berkaykrc** [comment]
> Let me know when PR is ready to get a review again 😄

Yeah sure, I just made the changes you requested, if that looks okay and no other adjustments are needed, we can merge after your review.

**Comment by berkaykrc** [comment]
> Hi @berkaykrc 🙂, I’ll try to review it as soon as possible. I recently changed jobs and I’m still trying to find the right balance.

Congrats on the new role! I can totally relate. I’m also in a transitional phase, currently in the final stage with a company and waiting for an offer. In the meantime, I’ve been diving into learning new things. No rush, just whenever you have the time.
